File: labs/lib/python/redaction.py
"""Shared redaction helpers for lab artifacts and HTTP traces."""

from __future__ import annotations

import re
from typing import Any

SENSITIVE_KEYS = frozenset(
    {
        "password",
        "passwd",
        "token",
        "authorization",
        "api_key",
        "api_secret",
        "api-key",
        "x-api-key",
        "client_secret",
        "client_id",
        "access_token",
        "refresh_token",
        "cookie",
        "set-cookie",
        "username",
        "partner_id",
        "signature",
        "secret",
        "portokasse",
        "application_id",
        "account_id",
        "account_number",
        "app_secret",
        "private_key",
    }
)

_SENSITIVE_FRAGMENTS = (
    "password",
    "passwd",
    "secret",
    "token",
    "authorization",
    "cookie",
    "apikey",
    "apisecret",
)

SECRET_KV_PATTERN = re.compile(
    r"(?i)\b(password|passwd|secret|token|api[_-]?key|client[_-]?secret)\b\s*([:=])\s*([^\s,;]+)"
)


def _normalized_key(key: str) -> str:
    return re.sub(r"[^a-z0-9]", "", key.lower())


_SENSITIVE_NORMALIZED = frozenset(_normalized_key(key) for key in SENSITIVE_KEYS)


def is_sensitive_key(key: str) -> bool:
    normalized = _normalized_key(key)
    if normalized in _SENSITIVE_NORMALIZED:
        return True
    return any(fragment in normalized for fragment in _SENSITIVE_FRAGMENTS)


def redact_headers(headers: dict[str, str] | None) -> dict[str, str]:
    if not headers:
        return {}
    out: dict[str, str] = {}
    for key, value in headers.items():
        out[key] = "[REDACTED]" if is_sensitive_key(key) else value
    return out


def redact_payload(payload: Any) -> Any:
    if isinstance(payload, dict):
        redacted: dict[str, Any] = {}
        for key, value in payload.items():
            if is_sensitive_key(str(key)):
                redacted[key] = "[REDACTED]"
            else:
                redacted[key] = redact_payload(value)
        return redacted
    if isinstance(payload, list):
        return [redact_payload(item) for item in payload]
    return payload


def redact_text(text: str) -> str:
    return SECRET_KV_PATTERN.sub(
        lambda match: f"{match.group(1)}{match.group(2)}[REDACTED]",
        text,
    )


def redact_url(url: str) -> str:
    from urllib.parse import parse_qsl, urlencode, urlsplit, urlunsplit

    parts = urlsplit(url)
    if not parts.query:
        return url
    redacted = []
    for key, value in parse_qsl(parts.query, keep_blank_values=True):
        if is_sensitive_key(key):
            redacted.append((key, "[REDACTED]"))
        else:
            redacted.append((key, value))
    return urlunsplit((parts.scheme, parts.netloc, parts.path, urlencode(redacted), parts.fragment))

File: labs/lib/typescript/redaction.ts
/** Shared redaction helpers for lab artifacts and HTTP traces. */

const SENSITIVE_KEYS = new Set([
    "password",
    "passwd",
    "token",
    "authorization",
    "api_key",
    "api_secret",
    "api-key",
    "x-api-key",
    "client_secret",
    "client_id",
    "access_token",
    "refresh_token",
    "cookie",
    "set-cookie",
    "username",
    "partner_id",
    "signature",
    "secret",
    "portokasse",
    "application_id",
    "account_id",
    "account_number",
    "app_secret",
    "private_key",
]);

const SENSITIVE_FRAGMENTS = [
    "password",
    "passwd",
    "secret",
    "token",
    "authorization",
    "cookie",
    "apikey",
    "apisecret",
];

const SECRET_KV_PATTERN =
    /\b(password|passwd|secret|token|api[_-]?key|client[_-]?secret)\b\s*([:=])\s*([^\s,;]+)/gi;

function normalizedKey(key: string): string {
    return key.toLowerCase().replace(/[^a-z0-9]/g, "");
}

const SENSITIVE_NORMALIZED = new Set([...SENSITIVE_KEYS].map(normalizedKey));

export function isSensitiveKey(key: string): boolean {
    const normalized = normalizedKey(key);
    if (SENSITIVE_NORMALIZED.has(normalized)) return true;
    return SENSITIVE_FRAGMENTS.some((fragment) => normalized.includes(fragment));
}

export function redactHeaders(headers: Record<string, string> | undefined): Record<string, string> {
    if (!headers) return {};
    const out: Record<string, string> = {};
    for (const [key, value] of Object.entries(headers)) {
        out[key] = isSensitiveKey(key) ? "[REDACTED]" : value;
    }
    return out;
}

export function redactPayload<T>(payload: T): T {
    if (payload === null || payload === undefined) return payload;
    if (Array.isArray(payload)) {
        return payload.map((item) => redactPayload(item)) as T;
    }
    if (typeof payload === "object") {
        const out: Record<string, unknown> = {};
        for (const [key, value] of Object.entries(payload as Record<string, unknown>)) {
            out[key] = isSensitiveKey(key) ? "[REDACTED]" : redactPayload(value);
        }
        return out as T;
    }
    return payload;
}

export function redactText(text: string): string {
    return text.replace(SECRET_KV_PATTERN, (_match, name: string, separator: string) => {
        return `${name}${separator}[REDACTED]`;
    });
}

export function redactUrl(url: string): string {
    try {
        const parsed = new URL(url);
        for (const key of [...parsed.searchParams.keys()]) {
            if (isSensitiveKey(key)) {
                parsed.searchParams.set(key, "[REDACTED]");
            }
        }
        return parsed.toString();
    } catch {
        return url;
    }
}
